cpp/src/umap/knn_graph/algo.cuh

```diff
@@ -101,6 +101,16 @@ inline void launcher(const raft::handle_t& handle,
                      const ML::UMAPParams* params,
                      cudaStream_t stream)
 {
+  size_t free_mem, total_mem;
+  cudaMemGetInfo(&free_mem, &total_mem);
+  double factor = 4.0;
+  size_t index_batch_size = inputsA.n;
+  size_t query_batch_size = inputsB.n;
+  size_t requirements = factor * sizeof(float) * index_batch_size * query_batch_size;
+  if (requirements > free_mem){
+    index_batch_size = free_mem / (query_batch_size * factor * sizeof(float));
+  }
```

Comment:
Fix looks good to me, can you just add a description of the fix to the PR description and fix the style issues?

Reply:
Sorry guys, this still isn't quite right.

Reply:
@cjnolet, this is an older comment regarding the previous version. The current version should make use of the `workspace_resource_factory` to create a `limiting_memory_resource` with the default amount (`total_size / 2 ?`) and adjust the index batch size according to available memory. There may totally be things I missed, but could you give more precision to what is required here?

Reply:
I updated the code slightly. But, since it's a RAFT call, the `limiting_memory_resource` cannot be used directly. The only thing that is doable is to update the handle, so that it has a resource factory that will in time be used to produce a workspace within RAFT sparse kNN code. We can produce an estimate of the memory that will be available on the cuML side by calling `resource::get_workspace_free_bytes` on the handle or `get_allocation_limit` on the LMR. But, that's about it, isn't it?

Reply:
@achirkin can you offer suggestions here? The main goal is to emable smart batching without relying on a query to get the free memory from the GPU (which is not acceptable practice). We started using the `limiting_memory_resource_adapter` so that we could do this while allowing the user to control this.

Reply:
I'm trying to understand the context here. In isolation, this change sets the batch size based on the available memory. If the `raft::sparse::selection::brute_force_knn` properly uses the workspace memory resource (which I hope it does but cannot check on my phone atm), then the solution looks good to me. Indeed, the workspace resource is always a limiting memory resource (the type is erased) and `resource::get_workspace_free_bytes` is a correct way to gets its remaining limit.

But I fail to see the connection to the linked issue. There, an invalid memory access error is reported, as suggested, likely due to integer overflow. If the workspace size is larger than 4gb, that still could be an issue, right?

Reply:
Thanks for taking a look @achirkin. The changes in https://github.com/rapidsai/raft/pull/1640 in addition to these changes should solve the issue. However, to allow any batch size to be used, it would be necessary to apply modifications to ensure that the integer values of `csr_batcher_t` and `distances_config_t` are also stored on 64 bits in order to guarantee that no operation would be made on 32 bits. This was much a more involved task, this is why I left it for later.